src/lib/posts/scaling-for-millions.md

```diff
@@ -0,0 +1,154 @@
+    <li>4. Load Balancer decides which server to route the request to</li>
+    <li>5. Server queries the database and recieves the response </li>
+    <li>6. Server sends the data back to Load Balancer</li>
+    <li>7. Load Balancer serves it back to user</li>
+</ol>
+</div>
+
+## Database Replication
+
+Winter is coming! so are Holidays. Are user-base is expanding like a balloon but we are ready with our scaled-out web-tier. The users will be served!
+One little thing to consider though. With all these users, our database has grown gaigantic too! At this point, if our one and only database goes down we will be in bit of a pickle. Also querying database for each and every request is expensive and can make the whole transaction super slow. To rectify this problem, we can replicate our database. Database replication is making one or more copies of the database which are constantly communicating with each other to keep up to date. This solves the problem of SPOF wherein if one of the database replicas is down, there are others to replace and serve on it's behalf.  For the expensive part, one of the database replicas is master replica which only hadles the write/update/delete operations while the other replicas(not master) are responsible for reading operations since read operations are more often. e.g. a write operation could be when a user wants to change their address while a read operation is user's information to be retrieved everytime they login/visit.
+
+*If you are wondering what happens when the Master replica goes down? Well one of the Not Master replicas will be elected as a new Master. Explaining it in detail will make this blog longer and I want to keep it short!*
+
+<div class="not-prose flex flex-col lg:flex-row justify-start items-center gap-10">
+<img src="/scaling/multi-server-with-db-replication.svg" alt="" width="800px"/>
+<ol class="flex flex-col max-w-[600px] gap-2 text-lg"> 
+    <li>1. User searches for our website </li>
+    <li>2. DNS resolves the name(domain) to its corresponding IP address which user will use to make request to the server </li>
+    <li>3. User makes request to the server that goes through LB </li>
+    <li>4. Load Balancer decides which server to route the request to</li>
+    <li>5. Server queries the read/write replica depending upon the nature of transaction </li>
+    <li>6. Database returns the data to server, server sends it LB</li>
+    <li>7. Load Balancer serves it back to user</li>
+</ol>
+</div>
+
+
+## Cache and CDN
+
+***Cache*** is a separate, relatively smaller storage system that stores frequently used data which helps in serving it blazing fast. When user reaches the server asking for his data, the server first goes on to check the cache and serves it right up if it's available. If it is not available in cache, it queries the database and serves it while also storing the data in cache for consequtive requests.
```

Comment:
@guacamole Please add some drama for `Blazingly fast`

Reply:
say no more